GVFS/GVFS/CommandLine/UpgradeVerb.cs

```diff
@@ -48,21 +48,23 @@ protected override string VerbName
         {
             get { return UpgradeVerbName; }
         }
-        
+
         public override void Execute()
         {
             ReturnCode exitCode = ReturnCode.Success;
-            if (!this.TryInitializeUpgrader() || !this.TryRunProductUpgrade())
+            string error;
+            if (!this.TryInitializeUpgrader(out error) || !this.TryRunProductUpgrade())

```

Comment:
At a high level, what happens if a user has downloaded an upgrader from GitHub, and then their organization switches to using a NuGet feed?

Will the upgrade verb somehow know that it shouldn't run the installer downloaded from GitHub, and that it should run the NuGet installer?

In general the question is how does the upgrade verb know which installer to run if GVFS.Service has downloaded different upgraders from different locations.  (If this is in the code and I've missed it, please point me in the right direction 😄 )

Reply:
Good question - the initial design does not do anything special if an organization switches to a different upgrader (e.g. switches to use a NuGet feed or vice versa). The upgrader does not uninstall any previous installations, it only runs the provided installer. I think this is reasonable for the initial version of this feature to keep complexity down, but I am interesting if there are opinions that this is insufficient.

In the current version of the NuGet upgrader change, If both upgraders are configured, the code currently prefers the NuGet upgrader. I could see this being an error condition (allowing the user to correct the ambiguous configuration).

Reply:
> The upgrader does not uninstall any previous installations, it only runs the provided installer. I think this is reasonable for the initial version of this feature to keep complexity down

I agree it's better to go with the less complex option for now.

> In the current version of the NuGet upgrader change, If both upgraders are configured, the code currently prefers the NuGet upgrader. I could see this being an error condition (allowing the user to correct the ambiguous configuration)

What will happen for organizations that use the current version of VFS4G (that's configured to download upgrades from GitHub) when they switch to using Nuget?

If an update was downloaded from GitHub and then they switch to Nuget will VFS4G do the right thing (assuming version numbers keep increasing)?  Or would the old upgrader downloaded from GitHub cause problems?

Reply:
>>If an update was downloaded from GitHub and then they switch to Nuget will VFS4G do the right thing (assuming version numbers keep increasing)? Or would the old upgrader downloaded from GitHub cause problems?

The old upgrader will not cause problems. 
1. Switching from GitHub to Nuget, while a previously downloaded GitHub release exists. Nuget upgrader downloads Nuget archive, un-zips it and uses the un-archived installer. Even if the older installer has the same name it will get overwritten (@jamill please correct me if I'm wrong.)
2. Switching from Nuget to GitHub, while a previously downloaded Nuget release exists. GitHubUpgrader always downloads new installer copy and uses it. It will overwrite any previously downloaded installers with the same name.